1_fetch/src/check_physio_region.R

```diff
@@ -0,0 +1,56 @@
+get_physio_regions <- function(reaches, ecoregions){
+  #' @description determines if a reach is within the physiographic ecoregion
+  #' based on reach intersection. This would ideally be completed by catchment 
+  #' areas within each ecoregion instead of reach intersection.
+  #'
+  #' @param reaches sf dataframe of PRMS reaches (lines)
+  #' @param ecoregions sf dataframe of ecoregions (multipolygon)
+  #' 
+  #' @return Returns reaches dataframe with one binary indicator columns
+  #' for each of the unique ecoregions.
+  
+  #Get a table of TRUE/FALSE for reach intersection with each ecoregion
+  physio_intersect <- st_intersects(x = reaches, y = ecoregions,
+                                    sparse = FALSE)
+  
+  #some reaches intersect several regions.
+  # assign binary indicator values in a table with columns as the region and row as reach
+  physio_mat <- data.frame(matrix(nrow = nrow(reaches), 
+                                  ncol = length(unique(ecoregions$Physio))))
+  colnames(physio_mat) <- sort(unique(ecoregions$Physio))
+  for(i in 1:ncol(physio_mat)){
+    #several polygons define each ecoregion, so using a function to check
+    # all of those polygons for each reach
+    physio_mat[,i] <- apply(physio_intersect, MARGIN = 1, 
+                            FUN = check_physio_region, region_name_check = colnames(physio_mat)[i],
+                            region_names = ecoregions$Physio) 
+  }
```

Comment:
> I'm curious if you know of a better way to extract the ecoregion data than what I programmed.

Woah, this approach is really clever! I don't know if I have a better way to extract the ecoregion data - do you want to know all of the ecoregions a line intersects, or just the one that overlaps the most? 

For spatial joins, I like the flexibility of `st_join(x, y, join = st_nearest_feature)` because you can choose a series of join operations. For example, I can specify `join = st_intersects` or `join = st_nearest_feature`. I'm guessing that's not so simple with this dataset though - `st_intersects` would return multiple lines if the reach crosses ecoregion boundaries; `st_nearest_feature` would (I think) find the ecoregion closest to one of the vertices so that ecoregion might not be the one that overlaps the most. 

If looking for the region that overlaps the most, I probably would approach it by finding the region with the greatest intersecting length:
```r
physio_intersections <- st_intersection(x = reaches, y = ecoregions) %>% 
    mutate(len = st_length(.)) %>%
    group_by(subsegid) %>%
    arrange(desc(len), .by_group = TRUE) %>%
    slice(1) %>%
    ungroup() %>%
    st_drop_geometry() %>%
    select(subsegid, Physio) 
  
reaches_w_regions <- left_join(x = reaches, y = physio_intersections, by = "subsegid")
head(reaches_w_regions, 2)
#> Simple feature collection with 2 features and 8 fields
#> Geometry type: LINESTRING
#> Dimension:     XY
#> Bounding box:  xmin: -74.61166 ymin: 42.14689 xmax: -74.46774 ymax: 42.356
#> Geodetic CRS:  WGS 84
#>   subsegid subsegseg subseglen fromsegs toseg tosubseg segidnat Physio                       geometry
#> 1      1_1         1  13588.58     <NA>     3      3_1     1435     AP LINESTRING (-74.51816 42.35...
#> 2      2_1         2  19134.73     <NA>     8      8_2     1436     AP LINESTRING (-74.47273 42.23...

```

That doesn't give you the binary matrix though. I tried modifying this quickly to keep all intersecting regions if that's what you're looking for. Hopefully these ideas are helpful for showing other ways to approach this problem. I haven't checked either of them for accuracy or against your original values. 

```r
physio_intersections <- st_intersection(x = reaches, y = ecoregions) %>%
  st_drop_geometry() %>%
  select(subsegid, Physio) %>%
  distinct() %>%
  pivot_wider(names_from = Physio, values_from = Physio) 
  
reaches_w_regions <- left_join(x = reaches, y = physio_intersections, by = "subsegid") %>%
  mutate(across(BR:VR, ~if_else(is.na(.), 0, 1)))

head(reaches_w_regions, 2)
#> Simple feature collection with 2 features and 12 fields
#> Geometry type: LINESTRING
#> Dimension:     XY
#> Bounding box:  xmin: -74.61166 ymin: 42.14689 xmax: -74.46774 ymax: 42.356
#> Geodetic CRS:  WGS 84
#>   subsegid subsegseg subseglen fromsegs toseg tosubseg segidnat BR AP CP PD VR                       geometry
#> 1      1_1         1  13588.58     <NA>     3      3_1     1435  0  1  0  0  0 LINESTRING (-74.51816 42.35...
#> 2      2_1         2  19134.73     <NA>     8      8_2     1436  1  1  0  0  0 LINESTRING (-74.47273 42.23...
```

Reply:
Thanks! This is helpful to see other approaches.

> do you want to know all of the ecoregions a line intersects, or just the one that overlaps the most?

Ideally, I'd want to know the ecoregion with the most drainage area overlap for the PRMS catchment, so instead of a binary matrix it would be a matrix of proportion drainage area within each region. I decided to keep it simple based on reach intersection instead of PRMS catchment intersection for time.